File: python/152_Maximum_Product_Subarray.py
#!/usr/bin/env python3

class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums or len(nums) == 0:
            return 0

        local_max = nums[0]
        local_min = nums[0]
        global_max = nums[0]

        for num in nums[1:]:
            copy_max = local_max
            local_max = max(num, copy_max * num, local_min * num)
            local_min = min(num, copy_max * num, local_min * num)

            if global_max < local_max:
                global_max = local_max

        return global_max


s = Solution()
print(s.maxProduct([2, 2, 0.4, 3]))

File: java/src/brand_new/_152_Maximum_Product_Subarray.java
package brand_new;

public class _152_Maximum_Product_Subarray {
	public int maxProduct(int[] nums) {
		if (nums.length == 0)
			return 0;

		int local_max = nums[0];
		int local_min = nums[0];
		int global_max = nums[0];
		
		for(int i = 1; i < nums.length; i++){
			int t = local_max;
			local_max = Math.max(nums[i], Math.max(local_max * nums[i], local_min * nums[i]));
			local_min = Math.min(nums[i], Math.min(t * nums[i], local_min * nums[i]));
			if(local_max > global_max)
				global_max = local_max;
		}
		return global_max;
	}

	public static void main(String[] args) {
		System.out.println(new _152_Maximum_Product_Subarray()
				.maxProduct(new int[] {-2,4,-2}));
	}
}
